src/lib/Export.hs

```diff
@@ -96,25 +95,45 @@ prepareFunctionForExport' cc f = do
             ety <- toExportType ty
             cont $ Nest (ExportResult (b:>ety)) Empty
 
-    exportArgRecon :: (EnvReader m, EnvExtender m) => Nest ExportArg n l -> m n (Abs (Nest IBinder) (ListE Block) n)
-    exportArgRecon topArgs = go [] topArgs
+    exportArgRecon
+      :: (EnvReader m, EnvExtender m)
+      => NaryPiType n
+      -> Nest ExportArg n l
+      -> m n (Abs (Nest IBinder) (ListE Block) n)
+    exportArgRecon (NaryPiType (NonEmptyNest tb tbs) _ _) topArgs =
+      go [] (Abs (Nest tb tbs) UnitE) topArgs
       where
         go :: (EnvReader m, EnvExtender m)
-           => [Block n] -> Nest ExportArg n l -> m n (Abs (Nest IBinder) (ListE Block) n)
-        go argRecons = \case
-          Empty -> return $ Abs Empty $ ListE argRecons
+           => [Block n]
+           -> Abs (Nest PiBinder) UnitE n
+           -> Nest ExportArg n l
+           -> m n (Abs (Nest IBinder) (ListE Block) n)
+        go argRecons dexArgs = \case
+          Empty -> case dexArgs of
+            Abs Empty UnitE -> return $ Abs Empty $ ListE argRecons
+            _ -> error "zip error: dex args should correspond 1-to-1 with export args"
           Nest (ExportArg _ b) bs ->
-            refreshAbs (Abs b (EmptyAbs bs)) \(v':>ety) (Abs bs' UnitE) -> do
-              (ity, block) <- typeRecon (sink ety) $ Var $ binderName v'
-              Abs ibs' allRecons' <- go (sinkList argRecons ++ [sink block]) bs'
-              return $ Abs (Nest (IBinder v' ity) ibs') allRecons'
-
-        typeRecon :: EnvExtender m => ExportType n -> Atom n -> m n (IType, Block n)
-        typeRecon ety v = case ety of
-          ScalarType sbt ->
-            return (Scalar sbt, Block (BlockAnn (BaseTy $ Scalar sbt) Pure) Empty v)
+            refreshAbs (Abs b (EmptyAbs bs)) \(v':>ety) (Abs bs' UnitE) ->
+              sinkM dexArgs >>= \case
+                Abs (Nest (PiBinder dexB dexArgTy _) restDexArgs) UnitE -> do
+                  -- get the current newtype-wrapped arg type
+                  -- pass it to type recon (who will use it to call `wrapWithNewtype` on the result
+                  -- to recur, wrap the var in the newtype and apply the subst
+                  (ity, block) <- typeRecon (sink ety) dexArgTy $ Var $ binderName v'
```

Comment:
Now that I look at it, I'm pretty sure that this always returns an atomic block, so you might be able to avoid the `wrapNewtypes` on the next line and just return the block result.

Reply:
I don't think I follow. I think `wrapNewtypes` is necessary for the substitution to be type-preserving

Reply:
but the result of the `block` is also type preserving, isn't it? And IMO it is always an `AtomicBlock` so you can just get it from there, instead of having to rewrap. In short, I'd try replacing `block` with `block@(AtomicBlock dexArgAtom)` here and seeing if that works

Reply:
I see! Yes, that works well. Just implemented it.